pkg/controlplane/controller/defaultservice/default_service_controller.go

```diff
@@ -0,0 +1,337 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package defaultservice
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"net/http"
+	"reflect"
+	"sync"
+	"time"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	genericapiserver "k8s.io/apiserver/pkg/server"
+	"k8s.io/apiserver/pkg/storage"
+	corev1informers "k8s.io/client-go/informers/core/v1"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/kubernetes/scheme"
+	v1core "k8s.io/client-go/kubernetes/typed/core/v1"
+	corev1listers "k8s.io/client-go/listers/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/klog/v2"
+	"k8s.io/kubernetes/pkg/controlplane/reconcilers"
+	netutils "k8s.io/utils/net"
+)
+
+const (
+	controllerName = "default-service-controller"
+	serviceName    = "kubernetes"
+)
+
+// New returns a new *Controller.
+func NewController(
+	serviceIP net.IP,
+	servicePort int,
+	serviceNodePort int,
+	endpointIP net.IP,
+	endpointPort int,
+	endpointsReconciler reconcilers.EndpointReconciler,
+	endpointInterval time.Duration,
+	endpointReconcilerType reconcilers.Type,
+	client clientset.Interface,
+) (*Controller, error) {
+	// The "kubernetes.default" Service is SingleStack based on the configured ServiceIPRange.
+	// If the bootstrap controller reconcile the kubernetes.default Service and Endpoints, it must
+	// guarantee that the Service ClusterIP and the associated Endpoints have the same IP family, or
+	// it will not work for clients because of the IP family mismatch.
+	// TODO: revisit for dual-stack https://github.com/kubernetes/enhancements/issues/2438
+	if endpointReconcilerType != reconcilers.NoneEndpointReconcilerType {
+		if netutils.IsIPv4(serviceIP) != netutils.IsIPv4(endpointIP) {
+			return nil, fmt.Errorf("service IP family %q must match endpoint (apiserver) address family %q", serviceIP.String(), endpointIP.String())
+		}
+	}
+
+	broadcaster := record.NewBroadcaster()
+	recorder := broadcaster.NewRecorder(scheme.Scheme, v1.EventSource{Component: controllerName})
+
+	c := &Controller{
+		endpointReconciler: endpointsReconciler,
+		serviceIP:          serviceIP,
+		servicePort:        servicePort,
+		serviceNodePort:    serviceNodePort,
+		endpointIP:         endpointIP,
+		endpointPort:       endpointPort,
+		client:             client,
+		endpointInterval:   endpointInterval,
+	}
+	// we construct our own informer because we need such a small subset of the information available.
+	// The kubernetes.default service
+	serviceInformer := corev1informers.NewFilteredServiceInformer(client, metav1.NamespaceDefault, 12*time.Hour,
+		cache.Indexers{cache.NamespaceIndex: cache.MetaNamespaceIndexFunc},
```

Comment:
Given we're watching from a single namespace (effectively a single object), creating an index here seems like overkill to me.